copy/copy.go

```diff
@@ -197,48 +207,110 @@ func AllowXAttrErrors(ci *CopyInfo) {
+	}
+}
+
 type copier struct {
-	chown             Chowner
-	utime             *time.Time
-	mode              *int
-	inodes            map[uint64]string
-	xattrErrorHandler XAttrErrorHandler
+	chown                 Chowner
+	utime                 *time.Time
+	mode                  *int
+	inodes                map[uint64]string
+	xattrErrorHandler     XAttrErrorHandler
+	includePatterns       []string
+	excludePatternMatcher *fileutils.PatternMatcher
 }
 
-func newCopier(chown Chowner, tm *time.Time, mode *int, xeh XAttrErrorHandler) *copier {
+func newCopier(chown Chowner, tm *time.Time, mode *int, xeh XAttrErrorHandler, includePatterns, excludePatterns []string) (*copier, error) {
 	if xeh == nil {
 		xeh = func(dst, src, key string, err error) error {
 			return err
 		}
 	}
-	return &copier{inodes: map[uint64]string{}, chown: chown, utime: tm, xattrErrorHandler: xeh, mode: mode}
+
+	var pm *fileutils.PatternMatcher
+	if len(excludePatterns) != 0 {
+		var err error
+		pm, err = fileutils.NewPatternMatcher(excludePatterns)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid excludepatterns: %s", excludePatterns)
+		}
+	}
+
+	return &copier{
+		inodes:                map[uint64]string{},
+		chown:                 chown,
+		utime:                 tm,
+		xattrErrorHandler:     xeh,
+		mode:                  mode,
+		includePatterns:       includePatterns,
+		excludePatternMatcher: pm,
+	}, nil
 }
 
 // dest is always clean
-func (c *copier) copy(ctx context.Context, src, target string, overwriteTargetMetadata bool) error {
+func (c *copier) copy(ctx context.Context, src, srcComponents, target string, overwriteTargetMetadata, includeAll bool) error {
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	default:
 	}
+
 	fi, err := os.Lstat(src)
 	if err != nil {
 		return errors.Wrapf(err, "failed to stat %s", src)
 	}
 
+	var include bool
+	if srcComponents != "" {
+		if !includeAll {
+			include, includeAll, err = c.include(srcComponents, fi)
+			if err != nil {
+				return err
+			}
+			if !include {
+				return nil
+			}
+		}
+		exclude, err := c.exclude(srcComponents, fi)
+		if err != nil {
+			return err
+		}
+		if exclude {
+			return nil
+		}
+	}
+
 	if !fi.IsDir() {
+		if include {
+			if err := c.createParentDirs(src, srcComponents, target, overwriteTargetMetadata); err != nil {
+				return err
+			}
+		}
 		if err := ensureEmptyFileTarget(target); err != nil {
 			return err
 		}
+	} else if includeAll {
+		if err := c.createParentDirs(src, srcComponents, target, overwriteTargetMetadata); err != nil {
+			return err
+		}
 	}
 
 	copyFileInfo := true
 
 	switch {
 	case fi.IsDir():
-		if created, err := c.copyDirectory(ctx, src, target, fi, overwriteTargetMetadata); err != nil {
+		if created, err := c.copyDirectory(ctx, src, srcComponents, target, fi, overwriteTargetMetadata, includeAll); err != nil {
 			return err
-		} else if !overwriteTargetMetadata {
+		} else if !overwriteTargetMetadata || !includeAll {
```

Comment:
bit confused about the includeAll check in here as exclude patterns seem to be not affected by this.

Reply:
`created` will always be false if `includeAll` is false (because in that case we delay creation of parent dirs). So this is shorter version of:

```go
if !includeAll {
        // directory may not have been created yet, so don't try to set its metadata
        copyFileInfo = false
}
```